dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java

```diff
@@ -360,29 +477,39 @@ void readBlock() throws HpccFileException
                 return;
             }
 
-            // Compact the array before reading in new data
-            int remainingBytesInBuffer = this.readBufferLen - this.readPos;
-            this.readBufferLen = remainingBytesInBuffer + dataLen;
+            while (dataLen > 0)
+            {
+                try
+                {
+                    this.bufferWriteMutex.acquire();
```

Comment:
do we know for sure we have enough room in the buffer at this point? Could we land here after sleeping (line 508)?

Reply:
We might not have enough to read the entire data block at this point in which case we will loop here until the consumer side reads enough data that we can read in the remaining data from the block